### ISSUE TITLE
uprobe support for JIT functions

### ISSUE THREAD

**Issue by mmarchini** [iovisor/bpftrace#425]
I wonder if we should add support for JIT functions to uprobes. It should be straightforward to implement by querying /tmp/perf-<pid>.map files.

That would allow us to trace dynamic languages functions.

If we decide to add this, should we look at the perf map only on startup, or should we watch the file for changes and attach to new probes when appropriate?

**Comment by brendangregg** [comment]
Some of us have already been working on it; we should chat in the next iovisor call about it.

**Comment by mmarchini** [comment]
Ok, this needs kernel work before we can move forward. uprobes are inode based and JIT functions are allocated in anonymous `mmap`ed pages during runtime. So we need to change the kernel to allow attaching to arbitrary executable memory on the process.

On the VM side it would be nice to stop moving code objects around while tracing the application.

**Comment by mmarchini** [comment]
I was looking at #790, and I think we might be able to use X watchpoints instead of uprobes for JIT functions. We could have a new probe type `jit:` if  we don't want to mix it with uprobes. Thoughts?

**Comment by brendangregg** [comment]
Good idea; although we need a way to pause the JIT compilers (JVM, v8) otherwise we'll end up watching the wrong thing.

**Comment by mmarchini** [comment]
> although we need a way to pause the JIT compilers (JVM, v8) otherwise we'll end up watching the wrong thing.

That would be ideal, but it's not a bpftrace issue (also, I think both JVM and V8 would get a huge performance hit from that, since they are built on the assumption that code will be optimized/deoptimized as more execution feedback is gathered).

**Comment by vchuravy** [comment]
I run into this a while back, hooking into the GDB mechanism mentioned earlier would be ideal since it would simplify the work needed for new languages. I wrote https://github.com/vchuravy/UProbes.jl to add uprobes support to Julia and currently have to go through the song and dance, where for each probe point I emit a shared-object file, dlopen that objectfile and then issue a foreign-function call from the probe location. 

Julia is a fully compiled JIT language and we don't to tracing optimization/deoptimization, but there might be multiple specialized copies of the same function that would resolve to the same probe name, but potentially with different arguments types.

Symbol resolution will be another challenge, as well as supporting functions being jitted after the process has already been instrumented.

**Comment by jpsamaroo** [comment]
I'm also interested in JIT support for Julia. @mmarchini has there been any progress in the kernel that you know of around using uprobes on `mmap`'d pages? Or is there an LKML discussion going on that we could follow?

**Comment by jrudolph** [comment]
Hwo 

> Ok, this needs kernel work before we can move forward. uprobes are inode based and JIT functions are allocated in anonymous `mmap`ed pages during runtime. So we need to change the kernel to allow attaching to arbitrary executable memory on the process.

Is it clear how that could work? I looked into https://github.com/torvalds/linux/blob/d20f6b3d747c36889b7ce75ee369182af3decb6b/kernel/events/uprobes.c#L1357 and it seems that uprobes a purely virtual memory based, i.e. there's a global hook that activates a probe whenever the right page for a given file would be mapped. How would that work for anonymous JIT regions?

> I think we might be able to use X watchpoints instead of uprobes for JIT functions

Is that something that would work with current kernels and is just missing plumbing?

**Comment by danobi** [comment]
bpftrace already has support for watchpoint probes on both userspace and kernelspace addresses. I don't use JITed languages much nor am I familiar with JIT internals. But if you give it a try and watchpoints don't work, just lemme know. I'll try to fix it.

FWIW I added watchpoint support a few years ago. I'm not sure there's much usage in the wild. But maybe it's a good fit for JIT.
